internal/kzg/kzg_prove.go

```diff
@@ -1,161 +0,0 @@
-//
-// We refer to q(X) as the quotient polynomial.
-//
-// The division needs to be handled differently if `z` is an element in the domain
-// as this means that one needs to divide by 0. Hence, you will observe that this function
-// will follow a different code-path depending on this condition.
-//
-// Note: Since compute f(z) necessitates one knowing whether `z` is in the domain,
-// this function accepts an `indexInDomain` value which will tell us the index of the
-// element if it is in the domain, see `evaluateLagrangePolynomial` for when `indexInDomain`
-// is computed.
-//
-// The matching code for this method is in `compute_kzg_proof_impl` where the quotient polynomial
-// is computed.
-func computeQuotientPoly(domain Domain, f Polynomial, indexInDomain int, fz, z fr.Element) ([]fr.Element, error) {
-	if domain.Cardinality != uint64(len(f)) {
-		return nil, ErrPolynomialMismatchedSizeDomain
-	}
-
-	if indexInDomain != -1 {
-		// Note: the uint64 conversion is both semantically correct and safer
-		// than accepting an `int``, since we know it shouldn't be negative
-		// and it should cause a panic, if not checked; uint64(-1) = 2^64 -1
-		return computeQuotientPolyOnDomain(domain, f, uint64(indexInDomain))
-	}
-
-	return computeQuotientPolyOutsideDomain(domain, f, fz, z)
-}
-
-// Computes q(X) = f(X) - f(z) / X - z in lagrange form where `z` is not in the domain.
-//
-// This function then performs division of two polynomials in evaluation form in the usual way.
-func computeQuotientPolyOutsideDomain(domain Domain, f Polynomial, fz, z fr.Element) ([]fr.Element, error) {
-
-	// Compute the lagrange form the of the numerator f(X) - f(z)
-	// Since f(X) is already in lagrange form, we can compute f(X) - f(z)
-	// by shifting all elements in f(X) by f(z)
-	numerator := make([]fr.Element, len(f))
-	for i := 0; i < len(f); i++ {
-		numerator[i].Sub(&f[i], &fz)
-	}
-
-	// Compute the lagrange form of the denominator X - z
-	// and invert all of the evaluations, since it is the
-	// denominator, we do a batch inversion.
-	denominator := make([]fr.Element, len(f))
-	for i := 0; i < len(f); i++ {
-		denominator[i].Sub(&domain.Roots[i], &z)
-	}
-	// Since `z` is not in the domain, we are sure that there
-	// are no zeroes in this inversion.
-	//
-	// Note: even if there was a zero, the gnark-crypto library would skip
-	// it and not panic.
-	denominator = fr.BatchInvert(denominator)
-
-	// Compute q(X)
-	for i := 0; i < len(f); i++ {
-		denominator[i].Mul(&denominator[i], &numerator[i])
-	}
-
-	return denominator, nil
-}
-
-// Computes f(X) - f(z) / X - z in lagrange form where `z` is in the domain.
-//
-// [compute_quotient_eval_within_domain](https://github.com/ethereum/consensus-specs/blob/3a2304981a3b820a22b518fe4859f4bba0ebc83b/specs/deneb/polynomial-commitments.md#compute_quotient_eval_within_domain)
-func computeQuotientPolyOnDomain(domain Domain, f Polynomial, index uint64) ([]fr.Element, error) {
-	fz := f[index]
```

Comment:
Nit: The way you are passing `indexInDomain` around seems correct (based on code review), but given that it goes through three different functions and also gets sign-casted, I wonder if we should also do a check here that `index < len(p)` or something.